### ISSUE TITLE
Support Gradient Boosted Trees

### ISSUE THREAD

**Issue by jonnor** [emlearn/emlearn#14]
Gradient boosted trees are very popular, and often show slightly better performance than RandomForest. scikit-learn has two implementations that we should support:

- sklearn.ensemble.GradientBoostingClassifier
- sklearn.ensemble.HistGradientBoostingClassifier

The HistGradientBoostingClassifier is new in scikit-learn 0.21 and implements more efficient training mechanism similar to what is used in LightGBM,xgboost,CatBoost. So maybe we don't need to implement those, at least right away?

Supporting gradient boosting trees will require implementing inference for regression trees (not just classification trees that we have now). So it should also make it possible implement regression:

- sklearn.ensemble.RandomForestRegressor
- sklearn.ensemble.ExtraTreesRegressor
- sklearn.ensemble.GradientBoostingRegressor
- sklearn.ensemble.HistGradientBoostingClassifier

**Comment by smokedlab** [comment]
Hi @jonnor , I’m interested in this issue. I have a background in computer engineering with a focus on embedded ML, so this looks very relevant to me.

I’d like to contribute, but I’m new to the codebase. I’ll start by going through how tree-based models are currently handled in emlearn. If it makes sense, I can begin with a small part (e.g. tests or a simple prototype for GradientBoostingClassifier).

Let me know if that sounds reasonable or if you have a preferred starting point.
cheers, Hadi

**Comment by jonnor** [comment]
Hi @smokedlab - nice to hear from another engineer focused on embedded ML :)

Your plan sounds good. I think that it might be best to add support for gradient-bosted trees as a separate submodule compared to decision trees/forests. Because the trees code is actually a bit complicated because it has grown organically over time, and have some semi-novel optimizations - that should not be the initial focus.
So create new file in src for a) the conversion code and b) for the C inference code. Maybe "gbtrees" is a decent name? But you can of course use the existing modules for inspiration.
Make sure to add a couple of simple test cases - again you can use that of trees as an example. Our primary testing strategy use scikit-learn as a reference in the tests. It is entirely fine to focus only on the most-common usecases first - such as classification. And you can start with just one of the "inference modes" - such as loadable. But I think the primary "interface" for classification should be predict_proba - returning probabilities. Especially for gradient-boosted-trees, since it naturally operates on probabilities. The default in scikit-learn of discretizing to single class index with predict() has several problems.

**Comment by smokedlab** [comment]
Hi @jonnor 

Thanks for the detailed input. I’ll take a closer look. This sounds like a solid plan, especially separating it into a dedicated module.

Appreciate the effort you put into outlining this. I might have a few questions once I dig into it and will get back to you if needed.
